File: tools/lib/jira.py
from __future__ import annotations

import base64
import json
from typing import Any

import requests


def _basic_auth(email: str, token: str) -> str:
    raw = f"{email}:{token}".encode("utf-8")
    return "Basic " + base64.b64encode(raw).decode("ascii")


def adf_to_text(node: Any) -> str:
    if node is None:
        return ""
    if isinstance(node, str):
        return node
    if isinstance(node, dict):
        parts: list[str] = []
        if node.get("type") == "text" and "text" in node:
            parts.append(str(node["text"]))
        for child in node.get("content") or []:
            parts.append(adf_to_text(child))
        return "".join(parts)
    if isinstance(node, list):
        return "".join(adf_to_text(x) for x in node)
    return ""


def fetch_issue(base_url: str, email: str, api_token: str, issue_key: str) -> dict[str, Any]:
    base = base_url.rstrip("/")
    url = f"{base}/rest/api/3/issue/{issue_key}"
    params = {
        "fields": "summary,description,labels,issuetype,status,priority,project",
    }
    r = requests.get(
        url,
        params=params,
        headers={"Accept": "application/json", "Authorization": _basic_auth(email, api_token)},
        timeout=60,
    )
    r.raise_for_status()
    return r.json()


def normalize_work_item(data: dict[str, Any]) -> dict[str, Any]:
    fields = data.get("fields") or {}
    desc = fields.get("description")
    desc_text = adf_to_text(desc) if isinstance(desc, dict) else (desc or "") or ""
    it = fields.get("issuetype") or {}
    st = fields.get("status") or {}
    pr = fields.get("priority") or {}
    proj = fields.get("project") or {}
    return {
        "provider": "jira",
        "key": data.get("key"),
        "summary": fields.get("summary") or "",
        "description": desc_text or None,
        "labels": list(fields.get("labels") or []),
        "issue_type": it.get("name"),
        "status": st.get("name"),
        "priority": pr.get("name"),
        "project_key": proj.get("key"),
        "raw": data,
    }


def handshake(base_url: str, email: str, api_token: str) -> dict[str, Any]:
    base = base_url.rstrip("/")
    url = f"{base}/rest/api/3/myself"
    r = requests.get(
        url,
        headers={"Accept": "application/json", "Authorization": _basic_auth(email, api_token)},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()

File: web/lib/jira.ts
import { normalizeJiraSiteUrl } from "./atlassian-urls";
import { abortAfter } from "./fetch-timeout";

export function basicAuthHeader(email: string, token: string): string {
  const raw = Buffer.from(`${email}:${token}`, "utf8").toString("base64");
  return `Basic ${raw}`;
}

export function adfToText(node: unknown): string {
  if (node == null) return "";
  if (typeof node === "string") return node;
  if (Array.isArray(node)) return node.map(adfToText).join("");
  if (typeof node === "object" && node !== null) {
    const o = node as Record<string, unknown>;
    const parts: string[] = [];
    if (o.type === "text" && typeof o.text === "string") parts.push(o.text);
    const content = o.content;
    if (Array.isArray(content)) parts.push(...content.map(adfToText));
    return parts.join("");
  }
  return "";
}

export async function fetchIssue(
  baseUrl: string,
  email: string,
  apiToken: string,
  issueKey: string
): Promise<Record<string, unknown>> {
  const base = normalizeJiraSiteUrl(baseUrl);
  const url = new URL(`${base}/rest/api/3/issue/${encodeURIComponent(issueKey)}`);
  url.searchParams.set("fields", "summary,description,labels,issuetype,status,priority,project");
  const r = await fetch(url.toString(), {
    headers: {
      Accept: "application/json",
      Authorization: basicAuthHeader(email, apiToken),
    },
    signal: abortAfter(60_000),
  });
  if (!r.ok) {
    const t = await r.text();
    throw new Error(`Jira ${r.status}: ${t.slice(0, 500)}`);
  }
  return r.json() as Promise<Record<string, unknown>>;
}

export function normalizeWorkItem(data: Record<string, unknown>): Record<string, unknown> {
  const fields = (data.fields as Record<string, unknown> | undefined) ?? {};
  const desc = fields.description;
  const descText =
    typeof desc === "object" && desc !== null ? adfToText(desc) : String(desc ?? "") || "";
  const it = fields.issuetype as { name?: string } | undefined;
  const st = fields.status as { name?: string } | undefined;
  const pr = fields.priority as { name?: string } | undefined;
  const proj = fields.project as { key?: string } | undefined;
  return {
    provider: "jira",
    key: data.key,
    summary: (fields.summary as string) ?? "",
    description: descText || null,
    labels: Array.isArray(fields.labels) ? [...(fields.labels as string[])] : [],
    issue_type: it?.name,
    status: st?.name,
    priority: pr?.name,
    project_key: proj?.key,
  };
}

export async function jiraHandshake(
  baseUrl: string,
  email: string,
  apiToken: string
): Promise<Record<string, unknown>> {
  const base = normalizeJiraSiteUrl(baseUrl);
  const r = await fetch(`${base}/rest/api/3/myself`, {
    headers: {
      Accept: "application/json",
      Authorization: basicAuthHeader(email, apiToken),
    },
    signal: abortAfter(30_000),
  });
  if (!r.ok) {
    const t = await r.text();
    throw new Error(`Jira ${r.status}: ${t.slice(0, 500)}`);
  }
  return r.json() as Promise<Record<string, unknown>>;
}
